app/percentage_controller.py

```diff
@@ -2,22 +2,31 @@
 import maya
 import cgi
 import qq
+from .models import User, Update, DBSession
 
 percentage = Blueprint('percentage', __name__)
 
 @percentage.route("/percentage", methods=['POST'])
 def percentage_update():
   data = request.get_json()
-  username = cgi.escape(data["username"])
+  zap_token = data["zap_token"]
+
   try:
     percentage = int(data["percentage"])
-    if percentage < 0 or percentage > 100 or len(username) > 16:
+    if percentage < 0 or percentage > 100 or len(zap_token) > 64:
       return "Reported percentage out of bounds", 400
 
     charging = bool(data["charging"])
     updated_at = maya.now().datetime(to_timezone='US/Eastern')
 
-    qq.insert_update(username, percentage, charging, updated_at)
+    with DBSession() as session:
+      user = session.query(User).filter(User.zap_token == zap_token).first()
+
+    if not user:
+      return "Zap token not recognized", 400
+
+    new_update = Update(user_id=user.id, percentage=percentage, charging=charging, updated_at=updated_at)
+    qq.insert_update(new_update, user.username)
```

Comment:
Gross as hell, any better way to get usernames for leaderboard? Prob should just call the damn DB

Reply:
yeah I am fixing this and just adding more DB calls, we can fix when we need performance